File: golang/leetcode/python/reserve_k_group.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reserveKGroup(self, head: ListNode, k: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        pre = dummy
        while pre.next:
            first = pre.next
            for i in range(k):
                if not first:
                    return dummy.next
                first = first.next
            second = first.next
            for i in range(k):
                if not first:
                    return dummy.next
                first = first.next
            pre.next = second
            first.next = second.next
            second.next = first
            pre = pre.next.next
        return dummy.next


"""
Test Case:
"""

def test_reserve_k_group():
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    head.next.next.next.next.next = ListNode(6)
    head.next.next.next.next.next.next = ListNode(7)
    head.next.next.next.next.next.next.next = ListNode(8)
    head.next.next.next.next.next.next.next.next = ListNode(9)
    head.next.next.next.next.next.next.next.next.next = ListNode(10)
    head.next.next.next.next.next.next.next.next.next.next = ListNode(11)
    head.next.next.next.next.next.next.next.next.next.next.next = ListNode(12)
    head.next.next.next.next.next.next.next.next.next.next.next.next = ListNode(13)
    head.next.next.next.next.next.next.next.next.next.next.next.next.next = ListNode(14)
    head.next.next.next.next.next.next.next.next.next.next.next.next.next.next = ListNode(15)
    head.next.next.next.next.next.next.next.next.next.next.next.next.next.next.next = ListNode(16)
    head.next.next.next.next.next.next.next.next.next.next.next.next.next.next.next.next = ListNode(17)
    Solution().reserveKGroup(head, 3)
    

File: golang/leetcode/js/reserve_k_group.js
const reserveKGroup = (head, k) => {
    if (head == null || k <= 1) return head;
    let dummy = new ListNode(0);
    dummy.next = head;
    let prev = dummy;
    let cur = head;
    let count = 0;
    while (cur) {
        let temp = cur;
        count = 0;
        while (temp && count < k) {
            temp = temp.next;
            count++;
        }
        if (count < k) break;
        let next = temp.next;
        temp.next = null;
        prev.next = reverse(cur);
        cur = next;
        prev = prev.next;
    }
    return dummy.next;
};

//test
let head = new ListNode(1);
head.next = new ListNode(2);
head.next.next = new ListNode(3);
head.next.next.next = new ListNode(4);
head.next.next.next.next = new ListNode(5);
console.log(reserveKGroup(head, 2));
console.log(reserveKGroup(head, 3));